src/psyclone/psyir/nodes/operation.py

```diff
@@ -224,38 +230,137 @@ def create(operator, lhs, rhs):
         binary_op.addchild(rhs)
         return binary_op
 
+    @property
+    def datatype(self):
+        '''
+        :returns: the datatype of the result of this BinaryOperation.
+        :rtype: :py:class:`psyclone.psyir.symbols.DataType`
+
+        :raises InternalError: if the types of the operands are not permitted
+                               for the given operator.
+        :raises NotImplementedError: if the operands are of the same intrinsic
+                                     type but different precisions.
+        :raises TypeError: if the operands are both arrays but are of different
+                           shapes.
+        '''
+        # Get the types of the operands.
+        argtypes = []
+        for child in self.children:
+            # If the operand is itself an operation this will recurse.
+            dtype = child.datatype
+            if isinstance(dtype, UnknownFortranType):
+                if dtype.partial_datatype:
+                    # We are still OK provided we have partial type information
+                    # since that means the intrinsic type can be handled in the
+                    # PSyIR.
+                    dtype = dtype.partial_datatype
+                else:
+                    return UnknownFortranType("")
+            if isinstance(dtype, DeferredType):
+                # If either operand is of DeferredType then we can't do
+                # any better.
+                return DeferredType()
+            if isinstance(dtype, ArrayType):
+                # We know this is an array but do we know its intrinsic type?
+                if isinstance(dtype.intrinsic, DeferredType):
+                    return DeferredType()
+                if isinstance(dtype.intrinsic, UnknownType):
+                    if (isinstance(dtype.intrinsic, UnknownFortranType) and
+                            dtype.intrinsic.partial_datatype):
+                        dtype = ArrayType(dtype.intrinsic.partial_datatype,
+                                          shape=dtype.shape)
+                    else:
+                        return UnknownFortranType("")
+
+            argtypes.append(dtype)
+
+        base_type = None
+
+        if self.operator not in self._numeric_ops:
+            # Must be a relational or logical operator. Intrinsic type of
+            # result will be boolean.
+            base_type = BOOLEAN_TYPE
+        else:
+            # We have a numerical operation.
+            if argtypes[0].intrinsic == argtypes[1].intrinsic:
+                if argtypes[0].precision != argtypes[1].precision:
+                    raise NotImplementedError(
+                        f"Cannot determine the type of expression "
+                        f"'{self.debug_string()}' involving arguments of the "
+                        f"same intrinsic type but different precision "
+                        f"('{argtypes[0].precision}' and "
+                        f"'{argtypes[1].precision}').")
```

Comment:
In Fortran'08 page 149 it says the expression type is that of the largest operand precision, is this something we want to check the the precisions are known, or we leave this for another PR?

Reply:
Ah! I didn't find that so thank you. I've extended the precision-handling support now.